### DISCUSSION TITLE
JPEG+RAW : Download ONLY jpeg file but save raw on memory card

### DISCUSSION THREAD

**Discussion by Akinetopsia1** [jim-easterbrook/python-gphoto2#156] [Q&A] [answered]
I've been attempting for the past few months to find a way to only download the jpeg for preview purposes while leaving/saving the raw file on the camera sdcard.

i basically want to replicate gphoto2's --keep-raw option like in

`gphoto2 --set-config capturetarget=1 --capture-image-and-download --keep-raw --frames 10 --interval 5`

If i use capturetarget_cfg.set_value('Memory card'), this code doesn't work and the files are only saved to the camera

```
    template = os.path.join(ROOT_DIR + WORK_DIR, 'frame%012d.jpg')
    timestamp = int( time.time() )

    path = camera.capture(gp.GP_CAPTURE_IMAGE)
    camera_file = camera.file_get(
        path.folder, path.name, gp.GP_FILE_TYPE_NORMAL)

    camera_file.save(template % timestamp)
```

Edit: I'm using the current python-gphoto2 v2.3.4 and latest libgphoto2 v2.5.30 / gphoto2 v2.5.28 and a Canon EOS m200

**Comment by jim-easterbrook** [comment]
I don't have that camera model but I expect you need to do the following:
1/ Set the camera to record RAW + JPEG
2/ Set the camera to record to memory card (I've just noticed that my camera reports that it records to internal RAM, even though it's actually using the memory card - I wonder if this setting is inverted somewhere).
3/ Use `camera.trigger_capture()` to take a photo.
4/ Use `camera.wait_for_event()` to wait for the jpeg to be saved.
5/ Transfer the jpeg to your computer and delete it from the camera.

**Comment by Akinetopsia1** [comment, marked as answer]
using camera.trigger_capture() instead of camera.capture() did not give any file added event. With a bit of fiddling and better error logging i managed to do what i wanted the following way, thanks to another discussion i found on here
```python
        #cleanup previous files
        test = os.listdir(ROOT_DIR + WORK_DIR)
        for item in test:
            if item.endswith(".jpg"):
                os.remove(os.path.join(ROOT_DIR + WORK_DIR, item))

        path = camera.capture(gp.GP_CAPTURE_IMAGE)
            
        #empty the event queue
        hasPreview = False
        typ,data = camera.wait_for_event(200)
        while typ != gp.GP_EVENT_TIMEOUT:

            if typ == gp.GP_EVENT_FILE_ADDED:
                if data.name.lower().endswith(".jpg"):
                    fn = os.path.join(data.folder,data.name)
                    messageQueue.append([ "camera_event", "text", "Downloading PREVIEW file: %s" % fn ])
                    camera_file = camera.file_get(
                        data.folder, data.name, gp.GP_FILE_TYPE_NORMAL)
                    camera_file.save(template % timestamp)
                    messageQueue.append([ "camera_event", "preview", ('frame%012d.jpg' % timestamp) ])
                    hasPreview = True
		
            #try to grab another event
            typ,data = camera.wait_for_event(1)
```

I'm not sure if waiting for gp_event_timeout is the proper way to go, but at least that works, albeit probably not in a very trustworthy way. Ie: can i be sure i'll always get the .jpg for the file added event? iirc previous tests i got the .cr3 instead. Time will tell i guess

Thx for the tip @jim-easterbrook, and thanks a bunch for the library wrapper, it's really nice to be able to mess with my camera for my projects!

**Reply by jim-easterbrook** [reply]
If the camera is saving raw and jpeg then there are two files generated for each capture. `camera.capture()` returns one of them, after waiting for it to be saved in the camera. `camera.trigger_capture()` returns immediately - useful if you want to capture on more than one camera at roughly the same time. The camera's events include all sorts of stuff we're not interested in, but need to be read before we get to the file added event that tells you when the jpeg or raw file has been stored in the camera.

**Reply by Akinetopsia1** [reply]
i might be doing something wrong with trigger_capture() then, or my late 2019 camera model is not quite acting like expected. Here's the events i get if i use that method and trace them using this lookup:

```python
    def event_text(event_type):
	if event_type == gp.GP_EVENT_CAPTURE_COMPLETE: return "Capture Complete"
	elif event_type == gp.GP_EVENT_FILE_ADDED: return "File Added"
	elif event_type == gp.GP_EVENT_FOLDER_ADDED: return "Folder Added"
	elif event_type == gp.GP_EVENT_TIMEOUT: return "Timeout"
	else: return "Unknown Event"

    #events are then logged in the while loop using this helper
    messageQueue.append([ "camera_event", "text", "Event: %s, data: %s" % (event_text(typ),data) ])
	
```

The first call i only get the following two events and then exits:

```
Event: Unknown Event, data: OLCInfo event 0x0010 content 04020a82
Event: Unknown Event, data: OLCInfo event mask=11
```


the following call returns more events but no file_added:

```
Event: Unknown Event, data: OLCInfo event mask=1
Event: Unknown Event, data: Camera Status 1
Event: Unknown Event, data: StorageInfoChanged 0x00020001
Event: Unknown Event, data: PTP Property d11b changed, "availableshots" to "2689"
Event: Unknown Event, data: OLCInfo event 0x0010 content 04020a81
Event: Unknown Event, data: OLCInfo event mask=10
```

Your comment also made me realize i had left my original call to camera.file_get() outside of the event loop -- removing it ~~now allows both raw and jpg file_added triggers within the wait_for_event loop~~, and actually makes the whole thing faster as i calling an extra file_get for no reason. I've edited my answer

edit: my bad. that is not correct. the initial call to camera.capture() returns the cr3 path info, and then i get the additional jpg later in the event loop. skipping get_file() for the raw file does save execution time though.

**Reply by jim-easterbrook** [reply]
Different cameras behave differently, and I don't have much experience with capturing. It may take some experimenting to find something that works for your setup. The `availableshots` change is a good sign though! I don't know how much timeout you need either.

**Reply by Akinetopsia1** [reply]
> Different cameras behave differently

Yeah, i came to that conclusion too from a few months of searching around.

> The availableshots change is a good sign

i still wouldn't know what the newly captured file names/paths are on the camera, tho unless there's a way to get that information otherwise
 
In the end I'm happy with my selected answer. With the extraneous call to get_file() removed, i get my preview on screen, and the raw left on the camera sd in about 3 seconds which is perfect

**Comment by Ayqi** [comment]
Hi, I'm trying the same workflow on my Nikon D5300 but I couldn't find a way to store the raw and jpeg files on the camera's sdcard. I would appreciate if you could share the complete capture code.
